pkg/services/librarypanels/librarypanels.go

```diff
@@ -39,6 +43,158 @@ func (lps *LibraryPanelService) IsEnabled() bool {
 	return lps.Cfg.IsPanelLibraryEnabled()
 }
 
+// LoadLibraryPanelsForDashboard loads library panels JSON from the database for a dashboard.
+func (lps *LibraryPanelService) LoadLibraryPanelsForDashboard(dash *models.Dashboard) error {
+	if !lps.IsEnabled() {
+		return nil
+	}
+
+	libraryPanels, err := lps.getLibraryPanelsForDashboardID(dash.Id)
+	if err != nil {
+		return err
+	}
+
+	panels := dash.Data.Get("panels").MustArray()
+	for i, panel := range panels {
+		panelAsJSON := simplejson.NewFromAny(panel)
+		libraryPanel := panelAsJSON.Get("libraryPanel")
+		if libraryPanel.Interface() == nil {
+			continue
+		}
+
+		// we have a library panel
+		uid := libraryPanel.Get("uid").MustString()
+		if len(uid) == 0 {
+			return errLibraryPanelHeaderUIDMissing
+		}
+
+		libraryPanelInDB, ok := libraryPanels[uid]
+		if !ok {
+			return fmt.Errorf("found connection to library panel %q that isn't in database", uid)
+		}
+
+		// we have a match between what is stored in db and in dashboard json
+		libraryPanelModel, err := libraryPanelInDB.Model.MarshalJSON()
+		if err != nil {
+			return fmt.Errorf("could not marshal library panel JSON: %w", err)
+		}
+
+		libraryPanelModelAsJSON, err := simplejson.NewJson(libraryPanelModel)
+		if err != nil {
+			return fmt.Errorf("could not convert library panel to simplejson model: %w", err)
+		}
+
+		// set the library panel json as new panel json in dashboard json
+		dash.Data.Get("panels").SetIndex(i, libraryPanelModelAsJSON.Interface())
+
+		// set dashboard specific props
+		elem := dash.Data.Get("panels").GetIndex(i)
+		elem.Set("gridPos", panelAsJSON.Get("gridPos").MustMap())
+		elem.Set("id", panelAsJSON.Get("id").MustInt64())
+		elem.Set("libraryPanel", map[string]interface{}{
+			"uid":  libraryPanelInDB.UID,
+			"name": libraryPanelInDB.Name,
+		})
+	}
+
+	return nil
+}
+
+// CleanLibraryPanelsForDashboard cleans library panels JSON before storing a dashboard to the database.
+func (lps *LibraryPanelService) CleanLibraryPanelsForDashboard(dash *models.Dashboard) error {
```

Comment:
Looking at the usage of this method and the tests, can't it just be called from `ConnectLibraryPanelsForDashboard` instead? Is there any reason to burden the client with calling both of these methods?

Reply:
When the first function is called, the dashboard doesn't have an Id/uid and we want to clean up the JSON that is stored before it's stored. That's why it's a 2 shot call. Does that make sense?

Reply:
@hugohaggmark OK, makes sense.

Reply:
To clarify, we can't connect the dashboard and library panels when there is no dashboardId and we don't want to store the whole library panel JSON in the db because that is already stored in the library_panel table.